### PR TITLE
.Net: Making ContextVariables into a proper Dictionary.

### PR DESCRIPTION
**Pull request by shawncal**
### Description
Simplifying ContextVariables by making it into an actual Dictionary<string, string> instead of a reimplemenation.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:
    - BREAKING CHANGE - use normal Dictionary methods instead. Behavior change: retrieving missing values via indexer [key] will throw if argument is missing, instead of returning string.Empty.

### CODE REVIEW

File: dotnet/src/IntegrationTests/TemplateLanguage/tests.txt [hunk 1]
```diff
@@ -18,9 +18,6 @@
 {{}}
 {{}}
 
-.{{asis}}.
-..
-
```
Comment (markwallace-microsoft) [inline dotnet/src/IntegrationTests/TemplateLanguage/tests.txt:23]: Why is this change required?

File: dotnet/src/IntegrationTests/Fakes/EmailSkillFake.cs [hunk 2]
```diff
@@ -11,7 +11,7 @@ internal sealed class EmailSkillFake
 {
     [SKFunction, Description("Given an email address and message body, send an email")]
     public Task<string> SendEmailAsync(
-        [Description("The body of the email message to send.")] string input,
+        [Description("The body of the email message to send.")] string input = "",
```
Comment (markwallace-microsoft) [inline dotnet/src/IntegrationTests/Fakes/EmailSkillFake.cs:14]: nit: Use `string.Empty`

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs [hunk 3]
```diff
@@ -21,15 +14,19 @@ namespace Microsoft.SemanticKernel.Orchestration;
...
     /// </summary>
     /// <param name="value">Optional value for the main variable of the context including trust information.</param>
     public ContextVariables(string? value = null)
+        : base(StringComparer.OrdinalIgnoreCase)
     {
-        this._variables[MainKey] = value ?? string.Empty;
+        if (value != null)
+        {
+            this[MainKey] = value;
+        }
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:29]: `MainKey` no longer defaults to an empty string, is that intentional?

If yes we should tag this as a breaking change as some clients may depend on this behaviour.

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs [hunk 4]
```diff
@@ -59,30 +56,7 @@ public ContextVariables Clone()
...
-            // If requested, discard old data and keep only the new one.
-            if (!merge) { this._variables.Clear(); }
-
-            foreach (KeyValuePair<string, string> varData in newData._variables)
-            {
-                this._variables[varData.Key] = varData.Value;
-            }
-        }
-
+        this.Set(MainKey, value);
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:59]: Allowing entry for `MainKey` to be set to null here but not in the constructor. Behaviour should be consistent.

File: dotnet/src/IntegrationTests/Fakes/EmailSkillFake.cs (hunk 2 above)
Comment (RogerBarreto) [reply dotnet/src/IntegrationTests/Fakes/EmailSkillFake.cs:14]: As this is a parameter we can't use `string.Empty` only const values.

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs (hunk 4 above)
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:59]: @markwallace-microsoft. According with the `Set` implementation, setting `null` will remove the variable.

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs (hunk 3 above)
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:29]: Call `Set(MainKey, value)` to ensure consistency.

```suggestion
        this.Set(MainKey, value);
```

File: dotnet/src/SemanticKernel.UnitTests/Orchestration/ContextVariablesTests.cs [hunk 5]
```diff
@@ -239,36 +239,6 @@ public void GetNameThatDoesNotExistReturnsFalse()
         Assert.Null(value);
     }
 
-    [Fact]
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.UnitTests/Orchestration/ContextVariablesTests.cs:242]: Keep the test. This should work for cloned Contexts as well.

**Review by RogerBarreto — CHANGES_REQUESTED** [review]

**Review by hario90 — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs [hunk 6]
```diff
@@ -21,15 +14,19 @@ namespace Microsoft.SemanticKernel.Orchestration;
 /// </summary>
 [DebuggerDisplay("{DebuggerDisplay,nq}")]
 [DebuggerTypeProxy(typeof(ContextVariables.TypeProxy))]
-public sealed class ContextVariables : IDictionary<string, string>
+public sealed class ContextVariables : Dictionary<string, string>
```
[the remark this answers is not in the document]
Comment (shawncal) [reply dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:17]: Possibly... let's work with object separately. We'll likely roll this in with ITypeConverter support

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs [hunk 7]
```diff
@@ -49,7 +46,7 @@ public ContextVariables Clone()
 
     /// <summary>Gets the main input string.</summary>
     /// <remarks>If the main input string was removed from the collection, an empty string will be returned.</remarks>
-    public string Input => this._variables.TryGetValue(MainKey, out string? value) ? value : string.Empty;
+    public string Input => this.TryGetValue(MainKey, out string? value) ? value : string.Empty;
```
[the remark this answers is not in the document]
Comment (shawncal) [reply dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs:49]: I agree. This just maintains current behavior, keeping this PR simple. I did try making `Input` nullable, and it results in a lot of changes to samples, tests.

This PR just seeks to just swap out the data structure with minimal breaking or functional changes.
We should decide between removing `Input` altogether (my preference), or keeping it and updating to nullable (and update callers). Either way, it's more readable if all of that is changed in another PR.

File: dotnet/src/IntegrationTests/Fakes/EmailSkillFake.cs [hunk 8]
```diff
@@ -20,8 +20,8 @@ internal sealed class EmailSkillFake
 
     [SKFunction, Description("Lookup an email address for a person given a name")]
     public Task<string> GetEmailAddressAsync(
-        [Description("The name of the person to email.")] string input,
-        ILogger logger)
+        ILogger logger,
+        [Description("The name of the person to email.")] string input = "")
```
[the remark this answers is not in the document]
Comment (shawncal) [reply dotnet/src/IntegrationTests/Fakes/EmailSkillFake.cs:24]: To maintain the logic in the method:
```
        if (string.IsNullOrEmpty(input))
        {
            logger.LogTrace("Returning hard coded email for {0}", input);
            return Task.FromResult("johndoe1234@example.com");
        }
```
If we want to allow no value to be passed, then we have to give it a default, as the property mapping uses defaults indicate optional parameters. `null` would fail to resolve from the dictionary and throw.
I'll update it to `string? input = null;` so that it's still optional, but not automatically empty.

File: dotnet/src/IntegrationTests/TemplateLanguage/tests.txt (hunk 1 above)
Comment (shawncal) [reply dotnet/src/IntegrationTests/TemplateLanguage/tests.txt:23]: The "asis" function has a required string input:
```
        [SKFunction, Description("This is a test"), SKName("asis")]
        public string MyFunction2(string input)
        {
            return input;
        }
```

Calling `asis` on its own is will crash because there's no input value provided. I can update asis to accept a nullable `string? input = null`

File: dotnet/src/SemanticKernel.UnitTests/Orchestration/ContextVariablesTests.cs (hunk 5 above)
Comment (shawncal) [reply dotnet/src/SemanticKernel.UnitTests/Orchestration/ContextVariablesTests.cs:242]: The `Update(ContextVariables)` method that this is testing has been removed.

I'll update the test so that it checks a similar scenario, but calls cv.Clone() instead.

File: dotnet/src/Skills/Skills.UnitTests/MsGraph/CalendarSkillTests.cs [hunk 9]
```diff
@@ -225,16 +225,13 @@ public async Task AddEventAsyncWithoutSubjectFailsAsync()
 
         CalendarSkill target = new(connectorMock.Object);
 
-        // Act
-        var ex = await Assert.ThrowsAsync<ArgumentException>(() => FunctionHelpers.CallViaKernel(target, "AddEvent",
+        // Act & Assert
+        var ex = await Assert.ThrowsAsync<SKException>(() => FunctionHelpers.CallViaKernel(target, "AddEvent",
```
Comment (lemillermicrosoft) [inline dotnet/src/Skills/Skills.UnitTests/MsGraph/CalendarSkillTests.cs:229]: nit: What exception is expected?

**Review by lemillermicrosoft — APPROVED** [review]

File: dotnet/src/IntegrationTests/TemplateLanguage/tests.txt (hunk 1 above)
Comment (shawncal) [reply dotnet/src/IntegrationTests/TemplateLanguage/tests.txt]: undone

File: dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs (hunk 3 above)
Comment (shawncal) [reply dotnet/src/SemanticKernel.Abstractions/Orchestration/ContextVariables.cs]: tagged as breaking

File: dotnet/src/Skills/Skills.UnitTests/MsGraph/CalendarSkillTests.cs (hunk 9 above)
Comment (shawncal) [reply dotnet/src/Skills/Skills.UnitTests/MsGraph/CalendarSkillTests.cs:229]: Missing arg throws:
SKException($"Missing value for parameter '{name}'");

(NativeFunction.cs:483)

Comment (shawncal) [reply dotnet/src/Skills/Skills.UnitTests/MsGraph/CalendarSkillTests.cs:230]: Added an additional check.

**Review by lemillermicrosoft — APPROVED** [review]

**Review by RogerBarreto — APPROVED** [review]

**Review by lemillermicrosoft — APPROVED** [review]
